File: Community/zo-ladder/scripts/unlock.py
#!/usr/bin/env python3
"""
Unlock Article - Paywall Bypass Tool
Fetches articles from paywalled sites and saves clean markdown.
"""

import argparse
import os
import re
import sys
import time
import urllib.parse
from pathlib import Path
from urllib.parse import urlparse

import requests
from readability import Document


def extract_domain(url):
    """Extract clean domain name from URL."""
    parsed = urlparse(url)
    domain = parsed.netloc.replace("www.", "").replace(".", "-")
    return domain


def extract_title_from_url(url):
    """Create a filename-friendly title from URL."""
    parsed = urlparse(url)
    path = parsed.path.strip("/")
    if path:
        # Get the last meaningful part of the path
        parts = [p for p in path.split("/") if p and not p.isdigit()]
        if parts:
            # Limit length and clean up
            title = parts[-1].replace("-", " ").replace("_", " ")[:50]
            return re.sub(r'[^\w\s-]', '', title).strip()
    return "article"


def try_archive_today(url):
    """Try to get article via archive.today."""
    try:
        encoded_url = urllib.parse.quote(url, safe="")
        submit_url = f"https://archive.today/submit/?url={encoded_url}"
        
        headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
        }
        
        # Submit URL to archive.today
        resp = requests.get(submit_url, headers=headers, allow_redirects=True, timeout=30)
        
        if resp.status_code == 200:
            # Return the archived page URL
            return resp.url
        return None
    except Exception as e:
        print(f"Archive.today error: {e}", file=sys.stderr)
        return None


def try_archive_org(url):
    """Try to get article via archive.org Wayback Machine."""
    try:
        # First check if there's a snapshot
        check_url = f"https://archive.org/wayback/available?url={urllib.parse.quote(url)}"
        resp = requests.get(check_url, timeout=10)
        
        if resp.status_code == 200:
            data = resp.json()
            if data.get("archived_snapshots"):
                snapshot = data["archived_snapshots"].get("closest")
                if snapshot and snapshot.get("available"):
                    return snapshot.get("url")
        
        # If no snapshot, try to save it
        save_url = f"https://web.archive.org/save/{url}"
        headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
        }
        resp = requests.get(save_url, headers=headers, allow_redirects=True, timeout=60)
        
        if resp.status_code == 200:
            return resp.url
        return None
    except Exception as e:
        print(f"Archive.org error: {e}", file=sys.stderr)
        return None


def try_textise_dot_iitty(url):
    """Try textise dot iitty service."""
    try:
        api_url = f"https://r.jina.ai/http://r.jina.ai/http://cc.bingj.com/cache.aspx?d=503-3904-1769&u={urllib.parse.quote(url)}"
        headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
        }
        
        resp = requests.get(api_url, headers=headers, timeout=30)
        if resp.status_code == 200 and len(resp.text) > 500:
            return resp.text
        return None
    except Exception as e:
        print(f"Textise error: {e}", file=sys.stderr)
        return None


def try_12ft_ladder(url):
    """Try 12ft.io ladder service."""
    try:
        ladder_url = f"https://12ft.io/{url}"
        headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
        }
        
        resp = requests.get(ladder_url, headers=headers, timeout=30)
        if resp.status_code == 200 and len(resp.text) > 500:
            return resp.text
        return None
    except Exception as e:
        print(f"12ft.io error: {e}", file=sys.stderr)
        return None


def extract_article_content(html, url):
    """Extract readable article content from HTML."""
    try:
        doc = Document(html)
        title = doc.title()
        content = doc.summary()
        
        # Clean up the content
        # Remove script and style tags
        import re
        content = re.sub(r'<script[^>]*>.*?</script>', '', content, flags=re.DOTALL)
        content = re.sub(r'<style[^>]*>.*?</style>', '', content, flags=re.DOTALL)
        
        # Convert to markdown-like format
        content = re.sub(r'<h1[^>]*>(.*?)</h1>', r'\n# \1\n', content, flags=re.DOTALL)
        content = re.sub(r'<h2[^>]*>(.*?)</h2>', r'\n## \1\n', content, flags=re.DOTALL)
        content = re.sub(r'<h3[^>]*>(.*?)</h3>', r'\n### \1\n', content, flags=re.DOTALL)
        content = re.sub(r'<p[^>]*>(.*?)</p>', r'\n\1\n', content, flags=re.DOTALL)
        content = re.sub(r'<li[^>]*>(.*?)</li>', r'- \1\n', content, flags=re.DOTALL)
        content = re.sub(r'<br\s*/?>', '\n', content)
        content = re.sub(r'<[^>]+>', '', content)
        
        # Clean up whitespace
        content = re.sub(r'\n{3,}', '\n\n', content)
        
        return title, content.strip()
    except Exception as e:
        print(f"Content extraction error: {e}", file=sys.stderr)
        return None, None


def save_article(title, content, url, output_dir):
    """Save article as markdown file."""
    domain = extract_domain(url)
    safe_title = re.sub(r'[^\w\s-]', '', title)[:60] if title else extract_title_from_url(url)
    safe_title = re.sub(r'\s+', '_', safe_title)
    
    # Create output directory
    article_dir = Path(output_dir) / domain
    article_dir.mkdir(parents=True, exist_ok=True)
    
    # Generate filename with timestamp to avoid collisions
    timestamp = time.strftime("%Y%m%d_%H%M%S")
    filename = f"{safe_title}_{timestamp}.md"
    filepath = article_dir / filename
    
    # Create markdown content
    markdown = f"""---
title: {title or 'Untitled Article'}
source_url: {url}
date_saved: {time.strftime("%Y-%m-%d %H:%M:%S")}
---

# {title or 'Untitled Article'}

*Source: [{url}]({url})*

---

{content}
"""
    
    with open(filepath, 'w', encoding='utf-8') as f:
        f.write(markdown)
    
    return filepath


def unlock_article(url, output_dir="/home/workspace/Articles"):
    """Main function to unlock and save an article."""
    print(f"Unlocking: {url}")
    
    html = None
    source = None
    
    # Try multiple methods
    methods = [
        ("12ft.io", try_12ft_ladder),
        ("archive.today", try_archive_today),
        ("archive.org", try_archive_org),
        ("textise service", try_textise_dot_iitty),
    ]
    
    for method_name, method_func in methods:
        print(f"  Trying {method_name}...")
        try:
            result = method_func(url)
            if result:
                if result.startswith("http"):
                    # It's a URL, fetch the content
                    headers = {"User-Agent": "Mozilla/5.0"}
                    resp = requests.get(result, headers=headers, timeout=30)
                    if resp.status_code == 200:
                        html = resp.text
                        source = method_name
                        break
                else:
                    # It's already HTML content
                    html = result
                    source = method_name
                    break
        except Exception as e:
            print(f"    {method_name} failed: {e}")
            continue
    
    if not html:
        print("ERROR: Could not bypass paywall with available methods.")
        print("This article may have a hard paywall or require authentication.")
        return None
    
    print(f"  Success! Retrieved via {source}")
    
    # Extract article content
    title, content = extract_article_content(html, url)
    
    if not content or len(content) < 200:
        print("WARNING: Extracted content is very short. The article may be truncated.")
    
    # Save the article
    filepath = save_article(title, content, url, output_dir)
    print(f"\nSaved to: {filepath}")
    print(f"Title: {title or 'Untitled'}")
    print(f"Content length: {len(content)} characters")
    
    return filepath


def main():
    parser = argparse.ArgumentParser(
        description="Unlock paywalled articles and save as markdown"
    )
    parser.add_argument("url", help="Article URL to unlock")
    parser.add_argument(
        "--output", "-o",
        default="/home/workspace/Articles",
        help="Output directory (default: /home/workspace/Articles)"
    )
    
    args = parser.parse_args()
    
    result = unlock_article(args.url, args.output)
    sys.exit(0 if result else 1)


if __name__ == "__main__":
    main()

File: Community/zo-ladder/scripts/unlock.ts
#!/usr/bin/env bun
/**
 * Unlock Article - Paywall Bypass Tool
 * Uses agent-browser for better paywall bypass capabilities
 */

import { $ } from "bun";

interface ArticleResult {
  success: boolean;
  filepath?: string;
  title?: string;
  error?: string;
}

function extractDomain(url: string): string {
  try {
    const parsed = new URL(url);
    return parsed.hostname.replace(/^www\./, "").replace(/\./g, "-");
  } catch {
    return "unknown";
  }
}

function sanitizeFilename(title: string): string {
  return title
    .replace(/[^\w\s-]/g, "")
    .replace(/\s+/g, "_")
    .slice(0, 60);
}

async function tryArchiveServices(url: string): Promise<string | null> {
  const services = [
    `https://r.jina.ai/http://${url}`,
    `https://r.jina.ai/http://cc.bingj.com/cache.aspx?d=503-3904-1769&u=${encodeURIComponent(url)}`,
  ];

  for (const serviceUrl of services) {
    try {
      const response = await fetch(serviceUrl, {
        headers: {
          "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
        },
      });

      if (response.ok) {
        const text = await response.text();
        if (text.length > 500) {
          return text;
        }
      }
    } catch (e) {
      // Continue to next service
    }
  }

  return null;
}

async function tryAgentBrowser(url: string): Promise<string | null> {
  try {
    // Use agent-browser to render the page fully
    const result = await $`agent-browser page ${url} --format markdown --wait 3000`.text();
    
    if (result.includes("security verification") || result.includes("CAPTCHA") || result.includes("bot")) {
      return null;
    }
    
    return result.length > 500 ? result : null;
  } catch {
    return null;
  }
}

async function tryZoReadWebpage(url: string): Promise<string | null> {
  try {
    // Try using read_webpage via Zo API
    const response = await fetch("https://api.zo.computer/zo/ask", {
      method: "POST",
      headers: {
        "Authorization": `Bearer ${process.env.ZO_CLIENT_IDENTITY_TOKEN || ""}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        input: `Read and extract the full article content from this URL: ${url}\n\nReturn the article title and full text content as markdown. If there's a paywall, try to bypass it using archive services or any available method.`,
      }),
    });

    if (response.ok) {
      const data = await response.json();
      return data.output;
    }
    return null;
  } catch {
    return null;
  }
}

async function extractWithReadability(html: string, url: string): Promise<{ title: string; content: string } | null> {
  try {
    // Use Mozilla's Readability via a simple extraction
    const titleMatch = html.match(/<title[^>]*>([^<]*)<\/title>/i);
    const title = titleMatch?.[1]?.trim() || "Untitled Article";

    // Basic content extraction - look for article content
    let content = html;

    // Remove script and style tags
    content = content.replace(/<script[^>]*>[\s\S]*?<\/script>/gi, "");
    content = content.replace(/<style[^>]*>[\s\S]*?<\/style>/gi, "");

    // Look for article or main content
    const articleMatch = content.match(/<article[^>]*>([\s\S]*?)<\/article>/i);
    const mainMatch = content.match(/<main[^>]*>([\s\S]*?)<\/main>/i);
    const contentMatch = content.match(/<div[^>]*class=["'][^"']*(?:content|article|post)[^"']*["'][^>]*>([\s\S]*?)<\/div>/i);

    let extractedContent = articleMatch?.[1] || mainMatch?.[1] || contentMatch?.[1] || content;

    // Convert to markdown
    extractedContent = extractedContent
      .replace(/<h1[^>]*>([\s\S]*?)<\/h1>/gi, "\n# $1\n")
      .replace(/<h2[^>]*>([\s\S]*?)<\/h2>/gi, "\n## $1\n")
      .replace(/<h3[^>]*>([\s\S]*?)<\/h3>/gi, "\n### $1\n")
      .replace(/<p[^>]*>([\s\S]*?)<\/p>/gi, "\n$1\n")
      .replace(/<li[^>]*>([\s\S]*?)<\/li>/gi, "- $1\n")
      .replace(/<br\s*\/?>/gi, "\n")
      .replace(/<[^>]+>/g, "");

    // Clean up
    extractedContent = extractedContent
      .replace(/&nbsp;/g, " ")
      .replace(/&amp;/g, "&")
      .replace(/&lt;/g, "<")
      .replace(/&gt;/g, ">")
      .replace(/\n{3,}/g, "\n\n")
      .trim();

    return { title, content: extractedContent };
  } catch {
    return null;
  }
}

async function saveArticle(
  title: string,
  content: string,
  url: string,
  outputDir: string
): Promise<string> {
  const domain = extractDomain(url);
  const safeTitle = sanitizeFilename(title);
  const timestamp = new Date().toISOString().replace(/[:.]/g, "-").slice(0, 19);
  const filename = `${safeTitle}_${timestamp}.md`;

  const articleDir = `${outputDir}/${domain}`;
  await $`mkdir -p ${articleDir}`;

  const filepath = `${articleDir}/${filename}`;

  const markdown = `---
title: ${title}
source_url: ${url}
date_saved: ${new Date().toISOString()}
---

# ${title}

*Source: [${url}](${url})*

---

${content}
`;

  await Bun.write(filepath, markdown);
  return filepath;
}

async function unlockArticle(url: string, outputDir: string): Promise<ArticleResult> {
  console.log(`Unlocking: ${url}`);

  let content: string | null = null;
  let source = "";

  // Try text extraction services first (fastest)
  console.log("  Trying text extraction services...");
  content = await tryArchiveServices(url);
  if (content) source = "extraction service";

  // Try agent-browser with wait
  if (!content) {
    console.log("  Trying browser automation...");
    content = await tryAgentBrowser(url);
    if (content) source = "browser";
  }

  // Try Zo API as last resort
  if (!content) {
    console.log("  Trying Zo webpage reader...");
    content = await tryZoReadWebpage(url);
    if (content) source = "Zo reader";
  }

  if (!content) {
    return {
      success: false,
      error: "Could not bypass paywall. This article may require authentication, have a hard paywall, or use strong bot protection.",
    };
  }

  console.log(`  Content retrieved via ${source}`);

  // Check if we got clean markdown directly
  let title: string;
  let articleContent: string;

  if (content.includes("#") && content.length > 1000) {
    // Looks like markdown already
    const titleMatch = content.match(/^#\s+(.+)$/m);
    title = titleMatch?.[1] || "Article";
    articleContent = content;
  } else {
    // Need to extract
    const extracted = await extractWithReadability(content, url);
    if (!extracted || extracted.content.length < 100) {
      return {
        success: false,
        error: "Could not extract readable article content from the retrieved page.",
      };
    }
    title = extracted.title;
    articleContent = extracted.content;
  }

  const filepath = await saveArticle(title, articleContent, url, outputDir);

  console.log(`\n✓ Article saved to: ${filepath}`);
  console.log(`  Title: ${title}`);
  console.log(`  Length: ${articleContent.length} characters`);

  return {
    success: true,
    filepath,
    title,
  };
}

// CLI
const url = process.argv[2];
const outputDir = process.argv[3] || "/home/workspace/Articles";

if (!url) {
  console.log("Usage: bun run unlock.ts <article-url> [output-directory]");
  process.exit(1);
}

if (!url.startsWith("http")) {
  console.log("Error: URL must start with http:// or https://");
  process.exit(1);
}

const result = await unlockArticle(url, outputDir);

if (!result.success) {
  console.error(`\n✗ ${result.error}`);
  process.exit(1);
}
